File: VB/View/CategoryView.vb
Imports DevExpress.XtraGrid.Views.Grid
Imports System.Collections.ObjectModel
Imports System.Collections.Specialized

Namespace DxExample
	Partial Public Class CategoryView
		Inherits UserControl

		Public Sub New()
			InitializeComponent()

			mvvmContext1.ViewModelType = GetType(CategoryViewModel)
			Dim fluentAPI = mvvmContext1.OfType(Of CategoryViewModel)()

			Dim expandedMasterRowsManager As New ExpandedMasterRowsManager(categoryGridView)

			fluentAPI.SetBinding(gridControl1, Function(grid) grid.DataSource, Function(x) x.Categories)

			fluentAPI.SetBinding(expandedMasterRowsManager, Function(x) x.ExpandedMasterRowsCollection, Function(y) y.ExpandedEntitiesCollection)
			fluentAPI.SetBinding(countLabelControl, Function(x) x.Text, Function(y) y.ExpandedEntitiesCount)

            fluentAPI.WithEvent(Of CustomMasterRowEventArgs)(categoryGridView, "MasterRowExpanded").EventToCommand(Sub(x) x.Update(Nothing), Function(ea) expandedMasterRowsManager.GetCategoryID(ea.RowHandle))
            fluentAPI.WithEvent(Of CustomMasterRowEventArgs)(categoryGridView, "MasterRowCollapsed").EventToCommand(Sub(x) x.Update(Nothing), Function(ea) expandedMasterRowsManager.GetCategoryID(ea.RowHandle))
            fluentAPI.WithEvent(categoryGridView, "DataSourceChanged").EventToCommand(Sub(x) x.UpdateDataBindings())

            fluentAPI.WithEvent(categoryIdEdit, "EditValueChanged").EventToCommand(Sub(x) x.UpdateCommands(GetCategoryIDFromEditor()))

            fluentAPI.BindCommand(buttonExpand, Sub(x) x.ExpandEntity(Nothing), Function(x) GetCategoryIDFromEditor())
            fluentAPI.BindCommand(buttonCollapse, Sub(x) x.CollapseEntity(Nothing), Function(x) GetCategoryIDFromEditor())
            fluentAPI.BindCommand(buttonDelete, Sub(x) x.DeleteEntity(Nothing), Function(x) GetCategoryIDFromEditor())
		End Sub

		Private Function GetCategoryIDFromEditor() As Integer
			Return Convert.ToInt32(categoryIdEdit.EditValue)
		End Function

		Private Class ExpandedMasterRowsManager
			Implements IDisposable

			Public Sub New(ByVal _view As GridView)
				view = _view
				SubscribeViewEvents(True)
			End Sub

			Private view As GridView
			Private _expandedMasterRowsCollection As ObservableCollection(Of Object)
			Private ReadOnly Property viewIsNull() As Boolean
				Get
					Return view Is Nothing
				End Get
			End Property
			Private Property lockExpandedMasterRowsCollectionChanged() As Boolean
			Private lockMasterRowExpanded As Boolean = False

			Public Property ExpandedMasterRowsCollection() As ObservableCollection(Of Object)
				Get
					Return _expandedMasterRowsCollection
				End Get
				Set(ByVal value As ObservableCollection(Of Object))
					If _expandedMasterRowsCollection Is value Then
						Return
					End If
					SubscribeCollectionChangedEvent(False)
					_expandedMasterRowsCollection = value
					SubscribeCollectionChangedEvent(True)
				End Set
			End Property

			Private Sub SubscribeCollectionChangedEvent(ByVal subscribe As Boolean)
				If _expandedMasterRowsCollection Is Nothing Then
					Return
				End If
				If subscribe Then
					AddHandler _expandedMasterRowsCollection.CollectionChanged, AddressOf ExpandedMasterRowsCollectionChanged
				Else
					RemoveHandler _expandedMasterRowsCollection.CollectionChanged, AddressOf ExpandedMasterRowsCollectionChanged
				End If
			End Sub

			Private Sub SubscribeViewEvents(ByVal subscribe As Boolean)
				If viewIsNull Then
					Return
				End If
				If subscribe Then
					AddHandler view.Disposed, AddressOf view_Disposed
					AddHandler view.MasterRowExpanded, AddressOf view_MasterRowExpanded
					AddHandler view.MasterRowCollapsed, AddressOf view_MasterRowCollapsed
					AddHandler view.DataSourceChanged, AddressOf view_DataSourceChanged
				Else
					RemoveHandler view.Disposed, AddressOf view_Disposed
					RemoveHandler view.MasterRowExpanded, AddressOf view_MasterRowExpanded
					RemoveHandler view.MasterRowCollapsed, AddressOf view_MasterRowCollapsed
					RemoveHandler view.DataSourceChanged, AddressOf view_DataSourceChanged
					view = Nothing
				End If
			End Sub

			Private Sub view_DataSourceChanged(ByVal sender As Object, ByVal e As EventArgs)
				lockExpandedMasterRowsCollectionChanged = True
				_expandedMasterRowsCollection.Clear()
				lockExpandedMasterRowsCollectionChanged = False
			End Sub

			Private Sub view_Disposed(ByVal sender As Object, ByVal e As EventArgs)
				SubscribeViewEvents(False)
			End Sub

			Private Sub view_MasterRowCollapsed(ByVal sender As Object, ByVal e As CustomMasterRowEventArgs)
				MasterRowExpanded(False, GetCategoryID(e.RowHandle))
			End Sub

			Private Sub view_MasterRowExpanded(ByVal sender As Object, ByVal e As CustomMasterRowEventArgs)
				MasterRowExpanded(True, GetCategoryID(e.RowHandle))
			End Sub

			Public Function GetCategoryID(ByVal rowHandle As Integer) As Integer
				Return (TryCast(view.GetRow(rowHandle), Category)).ID
			End Function

			Private Sub MasterRowExpanded(ByVal expanded As Boolean, ByVal categoryID As Integer)
				If lockMasterRowExpanded Then
					Return
				End If
				lockExpandedMasterRowsCollectionChanged = True
				If expanded Then
					_expandedMasterRowsCollection.Add(categoryID)
				Else
					_expandedMasterRowsCollection.Remove(categoryID)
				End If
				lockExpandedMasterRowsCollectionChanged = False
			End Sub

			Private Sub ExpandedMasterRowsCollectionChanged(ByVal sender As Object, ByVal e As NotifyCollectionChangedEventArgs)
				If lockExpandedMasterRowsCollectionChanged OrElse viewIsNull Then
					Return
				End If
				lockMasterRowExpanded = True
				Select Case e.Action
					Case NotifyCollectionChangedAction.Add
						For Each categoryID As Integer In e.NewItems
							ExpandMasterRow(categoryID)
						Next categoryID
					Case NotifyCollectionChangedAction.Remove
						For Each categoryID As Integer In e.OldItems
							CollapseMasterRow(categoryID)
						Next categoryID
					Case NotifyCollectionChangedAction.Reset
						CollapseAllMasterRows()
				End Select
				lockMasterRowExpanded = False
			End Sub

			Private Sub ExpandMasterRow(ByVal categoryID As Integer)
				view.ExpandMasterRow(GetRowHandle(categoryID))
			End Sub

			Private Sub CollapseMasterRow(ByVal categoryID As Integer)
				view.CollapseMasterRow(GetRowHandle(categoryID))
			End Sub

			Private Sub CollapseAllMasterRows()
				view.CollapseAllDetails()
			End Sub

			Private Function GetRowHandle(ByVal categoryID As Integer) As Integer
				Return view.LocateByValue("ID", categoryID, Nothing)
			End Function

			Public Sub Dispose() Implements IDisposable.Dispose
				SubscribeViewEvents(False)
				SubscribeCollectionChangedEvent(False)
			End Sub
		End Class
	End Class
End Namespace

File: CS/View/CategoryView.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows.Forms;
using DevExpress.XtraGrid.Views.Grid;
using System.Collections.ObjectModel;
using System.Collections.Specialized;

namespace DxExample
{
    public partial class CategoryView : UserControl
    {
        public CategoryView()
        {
            InitializeComponent();

            mvvmContext1.ViewModelType = typeof(CategoryViewModel);
            var fluentAPI = mvvmContext1.OfType<CategoryViewModel>();

            ExpandedMasterRowsManager expandedMasterRowsManager = new ExpandedMasterRowsManager(categoryGridView);

            fluentAPI.SetBinding(gridControl1, grid => grid.DataSource, x => x.Categories);

            fluentAPI.SetBinding(expandedMasterRowsManager, x => x.ExpandedMasterRowsCollection, y => y.ExpandedEntitiesCollection);
            fluentAPI.SetBinding(countLabelControl, x => x.Text, y => y.ExpandedEntitiesCount);

            fluentAPI.WithEvent<CustomMasterRowEventArgs>(categoryGridView, "MasterRowExpanded").EventToCommand(x => x.Update(null), ea => { return expandedMasterRowsManager.GetCategoryID(ea.RowHandle); });
            fluentAPI.WithEvent<CustomMasterRowEventArgs>(categoryGridView, "MasterRowCollapsed").EventToCommand(x => x.Update(null), ea => { return expandedMasterRowsManager.GetCategoryID(ea.RowHandle); });
            fluentAPI.WithEvent(categoryGridView, "DataSourceChanged").EventToCommand(x => x.UpdateDataBindings());

            fluentAPI.WithEvent(categoryIdEdit, "EditValueChanged").EventToCommand(x => x.UpdateCommands(GetCategoryIDFromEditor()));

            fluentAPI.BindCommand(buttonExpand, x => x.ExpandEntity(null), x => GetCategoryIDFromEditor());
            fluentAPI.BindCommand(buttonCollapse, x => x.CollapseEntity(null), x => GetCategoryIDFromEditor());
            fluentAPI.BindCommand(buttonDelete, x => x.DeleteEntity(null), x => GetCategoryIDFromEditor());
        }

        private int GetCategoryIDFromEditor()
        {
            return Convert.ToInt32(categoryIdEdit.EditValue);
        }

        class ExpandedMasterRowsManager : IDisposable
        {
            public ExpandedMasterRowsManager(GridView _view)
            {
                view = _view;
                SubscribeViewEvents(true);
            }

            GridView view;
            ObservableCollection<object> expandedMasterRowsCollection;
            bool viewIsNull { get { return view == null; } }
            bool lockExpandedMasterRowsCollectionChanged { get; set; }
            bool lockMasterRowExpanded = false;

            public ObservableCollection<object> ExpandedMasterRowsCollection
            {
                get
                {
                    return expandedMasterRowsCollection;
                }
                set
                {
                    if (expandedMasterRowsCollection == value) return;
                    SubscribeCollectionChangedEvent(false);
                    expandedMasterRowsCollection = value;
                    SubscribeCollectionChangedEvent(true);
                }
            }

            private void SubscribeCollectionChangedEvent(bool subscribe)
            {
                if (expandedMasterRowsCollection == null) return;
                if (subscribe)
                    expandedMasterRowsCollection.CollectionChanged += ExpandedMasterRowsCollectionChanged;
                else
                    expandedMasterRowsCollection.CollectionChanged -= ExpandedMasterRowsCollectionChanged;
            }

            private void SubscribeViewEvents(bool subscribe)
            {
                if (viewIsNull) return;
                if(subscribe)
                {
                    view.Disposed += view_Disposed;
                    view.MasterRowExpanded += view_MasterRowExpanded;
                    view.MasterRowCollapsed += view_MasterRowCollapsed;
                    view.DataSourceChanged += view_DataSourceChanged;
                }
                else
                {
                    view.Disposed -= view_Disposed;
                    view.MasterRowExpanded -= view_MasterRowExpanded;
                    view.MasterRowCollapsed -= view_MasterRowCollapsed;
                    view.DataSourceChanged -= view_DataSourceChanged;
                    view = null;
                }
            }

            private void view_DataSourceChanged(object sender, EventArgs e)
            {
                lockExpandedMasterRowsCollectionChanged = true;
                expandedMasterRowsCollection.Clear();
                lockExpandedMasterRowsCollectionChanged = false;
            }

            private void view_Disposed(object sender, EventArgs e)
            {
                SubscribeViewEvents(false);
            }

            private void view_MasterRowCollapsed(object sender, CustomMasterRowEventArgs e)
            {
                MasterRowExpanded(false, GetCategoryID(e.RowHandle));
            }

            private void view_MasterRowExpanded(object sender, CustomMasterRowEventArgs e)
            {
                MasterRowExpanded(true, GetCategoryID(e.RowHandle));
            }

            public int GetCategoryID(int rowHandle)
            {
                return (view.GetRow(rowHandle) as Category).ID;
            }

            private void MasterRowExpanded(bool expanded, int categoryID)
            {
                if (lockMasterRowExpanded) return;
                lockExpandedMasterRowsCollectionChanged = true;
                if (expanded)
                    expandedMasterRowsCollection.Add(categoryID);
                else
                    expandedMasterRowsCollection.Remove(categoryID);
                lockExpandedMasterRowsCollectionChanged = false;
            }

            private void ExpandedMasterRowsCollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
            {
                if (lockExpandedMasterRowsCollectionChanged || viewIsNull) return;
                lockMasterRowExpanded = true;
                switch (e.Action)
                {
                    case NotifyCollectionChangedAction.Add:
                        foreach (int categoryID in e.NewItems)
                            ExpandMasterRow(categoryID);
                        break;
                    case NotifyCollectionChangedAction.Remove:
                        foreach (int categoryID in e.OldItems)
                            CollapseMasterRow(categoryID);
                        break;
                    case NotifyCollectionChangedAction.Reset:
                        CollapseAllMasterRows();
                        break;
                }
                lockMasterRowExpanded = false;
            }

            private void ExpandMasterRow(int categoryID)
            {
                view.ExpandMasterRow(GetRowHandle(categoryID));
            }

            private void CollapseMasterRow(int categoryID)
            {
                view.CollapseMasterRow(GetRowHandle(categoryID));
            }

            private void CollapseAllMasterRows()
            {
                view.CollapseAllDetails();
            }

            private int GetRowHandle(int categoryID)
            {
                return view.LocateByValue("ID", categoryID, null); 
            }

            public void Dispose()
            {
                SubscribeViewEvents(false);
                SubscribeCollectionChangedEvent(false);
            }
        }
    }
}
